cmd/axelard/cmd/vald/evm/rpc/rpcClient.go

```diff
@@ -2,37 +2,100 @@ package rpc
 
 import (
 	"context"
+	"math/big"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	evmClient "github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
 )
 
 //go:generate moq -out ./mock/rpcClient.go -pkg mock . Client
 
+const codeMethodNotFound = -32601
+
 // Client provides calls to an EVM RPC endpoint
 type Client interface {
 	BlockNumber(ctx context.Context) (uint64, error)
 	TransactionByHash(ctx context.Context, hash common.Hash) (tx *types.Transaction, isPending bool, err error)
 	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+	BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error)
+	Close()
+
+	ChainGetFinalizedHead(ctx context.Context) (common.Hash, error)
+	ChainGetHeader(ctx context.Context, hash common.Hash) (*MoonbeamHeader, error)
+	IsMoonbeam() bool
 }
 
 // ClientImpl implements Client
 type ClientImpl struct {
 	*evmClient.Client
+	url        string
+	isMoonbeam bool
 }
 
-// NewClient returns an EVM rpc client
-func NewClient(url string) (*evmClient.Client, error) {
-	client, err := evmClient.Dial(url)
+// ChainGetFinalizedHead returns the hash of the latest finalized block
+func (c ClientImpl) ChainGetFinalizedHead(ctx context.Context) (common.Hash, error) {
+	rpc, err := rpc.DialContext(ctx, c.url)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	var result common.Hash
+	if err := rpc.CallContext(ctx, &result, "chain_getFinalizedHead"); err != nil {
+		return common.Hash{}, err
+	}
+
+	return result, nil
+}
+
+// ChainGetHeader returns the moonbeam block header of the given hash
+func (c ClientImpl) ChainGetHeader(ctx context.Context, hash common.Hash) (*MoonbeamHeader, error) {
+	rpc, err := rpc.DialContext(ctx, c.url)
 	if err != nil {
 		return nil, err
 	}
 
-	// try to access network
-	if _, err := client.ChainID(context.Background()); err != nil {
+	var result MoonbeamHeader
+	if err := rpc.CallContext(ctx, &result, "chain_getHeader", hash); err != nil {
 		return nil, err
 	}
 
-	return client, nil
+	return &result, nil
+}
+
+// IsMoonbeam returns true if the rpc client is connected to a moonbeam node; false otherwise
+func (c ClientImpl) IsMoonbeam() bool {
+	return c.isMoonbeam
+}
+
+// NewClient returns an EVM rpc client
+func NewClient(url string) (Client, error) {
+	c, err := evmClient.Dial(url)
+	if err != nil {
+		return nil, err
+	}
+
+	client := ClientImpl{
+		Client: c,
+		url:    url,
+	}
+
+	_, err = client.ChainGetFinalizedHead(context.Background())
+	switch err := err.(type) {
+	case nil:
+		client.isMoonbeam = true
```

Comment:
It's currently moonbeam only and these RPCs are no standard whatsoever. It's better to be GRANDPA instead of moonbeam but we are not sure if we are gonna deal with any GRANDPA chain in the future and even if we do we don't know if they are gonna implement these non-standard JSON-RPCs.

Reply:
If there's some other chain later supports a non-standard JSON-RPC like `chain_getFinalizedHead`, it's highly likely they are a GRANDPA chain and the finality process remains working. In the highly unlikely case for a proof-of-work evm chain to support `chain_getFinalizedHead`, as long as this method still means the same thing, then we also won't have any problem with it.